File: server/src/test/java/io/microconfig/server/git/GitServiceIT.java
package io.microconfig.server.git;

import org.junit.Test;

import java.io.File;

import static org.assertj.core.api.Assertions.assertThat;

public class GitServiceIT {
    private File local = new File(System.getProperty("user.home") + "/microconfig");

    @Test
    public void should_checkout_default() {
        var config = gitConfig("https://github.com/microconfig/configs-layout-example.git");

        //when
        var git = new GitServiceImpl(config);
        git.checkoutDefault();

        //then
        testDir("configs-layout-example");
    }

    @Test
    public void should_checkout_open() {
        var config = gitConfig("https://github.com/microconfig/configs-layout-example.git");

        //when
        var git = new GitServiceImpl(config);
        git.checkoutBranch("vault");

        //then
        testDir("configs-layout-example");
    }

    @Test
    public void should_checkout_tag() {
        var config = gitConfig("https://github.com/microconfig/configs-layout-example.git");

        //when
        var git = new GitServiceImpl(config);
        git.checkoutTag("vault-tag");

        //then
        testDir("configs-layout-example");
    }

    @Test
    public void should_checkout_private() {
        var config = gitConfig("private url");
        config.setUsername("user");
        config.setPassword("pass");

        //when
        var git = new GitServiceImpl(config);
        git.checkoutDefault();

        //then
        testDir("name");
    }

    private GitConfig gitConfig(String remote) {
        var config = new GitConfig();
        config.setRemoteUrl(remote);
        config.setWorkingDir(local);
        return config;
    }

    private void testDir(String dirName) {
        var expected = new File(local, dirName);
        assertThat(expected).exists();
        assertThat(expected).isDirectory();
    }
}

File: server/src/test/kotlin/io/microconfig/server/git/GitServiceIT.kt
package io.microconfig.server.git

import io.microconfig.server.git.exceptions.RefNotFound
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.io.File

class GitServiceIT {
    private val local = File(System.getProperty("user.home") + "/microconfig")
    private val repo = "https://github.com/microconfig/microconfig-quickstart.git"

    @Test
    fun should_checkout_default() {
        val config = gitConfig(repo)

        //when
        val git = GitServiceImpl(config)
        git.checkoutRef(null)

        //then
        testDir(config, "master")
    }

    @Test
    fun should_checkout_branch() {
        val config = gitConfig(repo)

        //when
        val git = GitServiceImpl(config)
        git.checkoutRef("vault")

        //then
        testDir(config, "vault")
    }

    @Test
    fun should_checkout_tag() {
        val config = gitConfig("https://github.com/microconfig/microconfig-quickstart.git")

        //when
        val git = GitServiceImpl(config)
        git.checkoutRef("test-tag")

        //then
        testDir(config, "test-tag")
    }

    @Test
    fun should_fail_on_missing_tag() {
        val config = gitConfig("https://github.com/microconfig/microconfig-quickstart.git")
        val git = GitServiceImpl(config)

        //expected
        assertThrows<RefNotFound> { git.checkoutRef("missing-tag") }
        val expected = File(config.dir(), "missing-tag")
        assertThat(expected).doesNotExist()
    }

    //    @Test
    fun should_checkout_private() {
        val config = gitConfig("private url")
        config.username = "user"
        config.password = "pass"

        //when
        val git = GitServiceImpl(config)
        git.checkoutRef("master")

        //then
        testDir(config, "name")
    }

    private fun gitConfig(remote: String): GitConfig {
        return GitConfig(local, remote)
    }

    private fun testDir(config: GitConfig, dirName: String) {
        val expected = File(config.dir(), dirName)
        assertThat(expected).exists()
        assertThat(expected).isDirectory
    }
}